### PR TITLE
Add support for indirect drawing

### PR DESCRIPTION
**Pull request by dragly**

### CODE REVIEW

**Comment by kvark** [comment]
@dragly thank you for the PR! Is it still WIP? I think we can land it :)

**Comment by dragly** [comment]
Thanks! I just wanted to test that `draw_indirect` worked before submitting it as final. I had only tested the indexed version when I pushed this. I will try to find time to add a PR with a couple of examples to wgpu-rs that makes use of the new functions.

Also, since I am not familiar with the entire wgpu-native and gfx pipelines, I was not 100 % sure what was needed for the implementation, such as if calling `pass.is_ready().unwrap()` is necessary. I just went ahead with an attempt based on the other draw-functions. Please double check and let me know if I have missed anything :)

**Comment by kvark** [comment]
All good. Thank you!
bors r+

File: wgpu-native/src/resource.rs [hunk 1]
```diff
@@ -162,6 +162,7 @@ bitflags! {
         const SAMPLED = 4;
         const STORAGE = 8;
         const OUTPUT_ATTACHMENT = 16;
+        const INPUT_ATTACHMENT = 32;
```
Comment (kvark) [inline wgpu-native/src/resource.rs]: this is not a valid usage in WebGPU

File: wgpu-native/src/command/compute.rs [hunk 2]
```diff
@@ -158,6 +160,29 @@ pub extern "C" fn wgpu_compute_pass_dispatch(pass_id: ComputePassId, x: u32, y:
...
+    let mut token = Token::root();
+    let (buffer_guard, _) = HUB.buffers.read(&mut token);
+    let (mut pass_guard, _) = HUB.compute_passes.write(&mut token);
+    let pass = &mut pass_guard[pass_id];
+
+    let buffer = pass
+        .trackers
+        .buffers
+        // TODO figure out why STORAGE was needed here and not INDIRECT
+        .use_extend(&*buffer_guard, indirect_buffer_id, (), BufferUsage::STORAGE)
```
Comment (kvark) [inline wgpu-native/src/command/compute.rs]: it should definitely be `INDIRECT`

**Comment by kvark** [comment]
Thank you for updating the patch!
I noted my concerns with the new code above. Also, please rebase instead of merging:
> Merge branch 'master' into dragly/indirect

**Comment by dragly** [comment]
Ah, sorry. I forgot this PR was still open as I thought it was already merged. The indirect dispatch was not intended to go together with indirect drawing. I should have pushed this to a different branch. I will update this to only involve drawing and push a second PR.

Where is the commit that bors references this PR from, by the way? I could not find it in the logs anywhere.

**Comment by kvark** [comment]
> Where is the commit that bors references this PR from, by the way? I could not find it in the logs anywhere.

I think it's in the staging (or "staging.tmp"?) branch:
https://github.com/gfx-rs/wgpu/commit/2729493d0c3d9f393936e9778c3801c94418b81d

It shouldn't matter though. You should only make sure your commits are good, not the ones Bors does :)

**Comment by dragly** [comment]
I changed this branch to only include the indirect drawing. 

But I do not get how this merging and syncing works via mozilla central. The indirect drawing commands are still missing from render.rs on both master and staging, and I cannot find the commit in master or staging (and there is no staging.tmp branch). Is this why this PR is still open?

**Comment by kvark** [comment]
> But I do not get how this merging and syncing works via mozilla central

It's not yet in place, anyway. You are confused by something else here.

**Comment by dragly** [comment]
Oh, okay. Then I just misunderstood. I see it got merged now. Thanks!

**Pull request merged**